redun/executors/gcp_batch.py

```diff
@@ -0,0 +1,584 @@
+            script_command = ["bash", script_path]
+            gcp_job = gcp_utils.batch_submit(
+                client=self.gcp_client,
+                job_name=f"redun-{job.id}",
+                project=project,
+                region=region,
+                image=image,
+                commands=script_command,
+                mount_buckets=mount_buckets,
+                **task_options,
+            )
+
+        job_dir = get_job_scratch_dir(self.gcs_scratch_prefix, job)
+        self.log(
+            "submit redun job {redun_job} as {job_type}:\n"
+            "  task_name          = {task_name}\n"
+            "  gcs_scratch_path = {job_dir}\n".format(
+                redun_job=job.id,
+                job_type="GCP Batch Job Task",
+                task_name=gcp_job.task_groups[0].name,
+                job_dir=job_dir,
+            )
+        )
+        task_name = gcp_job.task_groups[0].name
+        # For a single submitted task we can set task name directly
+        self.pending_batch_tasks[f"{task_name}/tasks/0"] = job
+
+    def _start(self) -> None:
+        """
+        Start monitoring thread.
+        """
+        if not self._thread or not self._thread.is_alive():
+            self.is_running = True
+            self._thread = threading.Thread(target=self._monitor, daemon=False)
+            self._thread.start()
+
+    def stop(self) -> None:
+        """
+        Stop Executor and monitoring thread.
+        """
+        self.is_running = False
+
+        self._docker_executor.stop()
+        self.arrayer.stop()
+
+        # Stop monitor thread.
+        if (
+            self._thread
+            and self._thread.is_alive()
+            and threading.get_ident() != self._thread.ident
+        ):
+            self._thread.join()
+
+    def _monitor(self) -> None:
+        """
+        Thread for monitoring running GCP Batch jobs.
+        """
+        assert self._scheduler
+
+        # Need new client for thread safety
+        gcp_client = gcp_utils.get_gcp_client()
+
+        try:
+            while self.is_running and (self.pending_batch_tasks or self.arrayer.num_pending):
+                if self._scheduler.logger.level >= logging.DEBUG:
+                    self.log(
+                        f"Preparing {self.arrayer.num_pending} job(s) for Job Arrays.",
+                        level=logging.DEBUG,
+                    )
+                    self.log(
+                        f"Waiting on {len(self.pending_batch_tasks)} Batch tasks(s):\n\t"
+                        + "\n\t".join(sorted(self.pending_batch_tasks.keys())),
+                        level=logging.DEBUG,
+                    )
+
+                # Copy pending_batch_tasks.keys() since it can change due to new submissions.
+                task_names = list(self.pending_batch_tasks.keys())
+                for name in task_names:
+                    try:
+                        task = gcp_utils.get_task(client=gcp_client, task_name=name)
```

Comment:
Is there a bulk status API that can be used? Polling each job individually, might end up being a lot API calls for large workflows.

Reply:
Yes this could definitely end up being a lot of API calls... but I don't think there is a bulk status API. I will double check with the batch folks.

But there is a method to configure a GCP batch job to notify a GCP Pub/Sub topic status changes. Then, on the redun side we can subscribe to the topic and receive changes as they occur instead of polling. That might be a nice addition?